module/zfs/dbuf.c

```diff
@@ -683,6 +683,18 @@ dbuf_evict_one(void)
 	    multilist_sublist_t *, mls);
 
 	if (db != NULL) {
+		/*
+		 * Bail out if another process is already evicting
+		 * this dbuf's objset.
+		 */
+		uint64_t tid = (uint64_t)(uintptr_t)curthread;
+		objset_t *os = db->db_objset;
+		(void) atomic_cas_64(&os->os_ev_tid, 0, tid);
+		if (os->os_ev_tid != tid) {
+			multilist_sublist_unlock(mls);
+			mutex_exit(&db->db_mtx);
+			return;
+		}
```

Comment:
Could you explain the deadlock your resolving with this change (stack traces would be great).  I'd like to better understand what's wrong with the existing locking.

Reply:
{code}
root 2833 2 0 16:05 ? 00:00:00 [dbu_evict]
[<ffffffffc069d46b>] spl_panic+0xeb/0x100 [spl]
[<ffffffffc03e7084>] avl_insert_here+0x3d4/0x410 [zavl]
[<ffffffffc0839de7>] dnode_evict_dbufs+0x1f7/0x280 [zfs]
[<ffffffffc081b816>] dmu_objset_evict_dbufs+0x136/0x220 [zfs]
[<ffffffffc081f7ae>] dmu_objset_evict+0x5e/0x170 [zfs]
[<ffffffffc083df5b>] dsl_dataset_evict_async+0x3b/0x280 [zfs]
[<ffffffffc06a5bb2>] taskq_thread+0x262/0x4d0 [spl]
[<ffffffff8eea8de9>] kthread+0x109/0x140
[<ffffffff8f6d413c>] ret_from_fork+0x2c/0x40
[<ffffffffffffffff>] 0xffffffffffffffff
root 2834 2 0 16:05 ? 00:00:02 [dbuf_evict]
[<ffffffffc0803a5b>] dbuf_destroy+0x1fb/0x6f0 [zfs]
[<ffffffffc0804061>] dbuf_evict_one+0x111/0x1c0 [zfs]
[<ffffffffc0804251>] dbuf_evict_thread+0x141/0x270 [zfs]
[<ffffffffc06a6e4b>] thread_generic_wrapper+0x7b/0xc0 [spl]
[<ffffffff8eea8de9>] kthread+0x109/0x140
[<ffffffff8f6d413c>] ret_from_fork+0x2c/0x40
[<ffffffffffffffff>] 0xffffffffffffffff
root 18742 17666 0 16:27 pts/18 00:00:00 /root/snapexists/zfs/cmd/zpool/.libs/lt-zpool destroy -f testpool1
[<ffffffffc06a48e4>] taskq_wait+0x64/0xa0 [spl]
[<ffffffffc080bbc5>] dmu_buf_user_evict_wait+0x15/0x20 [zfs]
[<ffffffffc08a1cab>] spa_evicting_os_wait+0x9b/0xd0 [zfs]
[<ffffffffc0895840>] spa_export_common+0x3f0/0x510 [zfs]
[<ffffffffc089597a>] spa_destroy+0x1a/0x20 [zfs]
[<ffffffffc08fcc7a>] zfs_ioc_pool_destroy+0x1a/0x20 [zfs]
[<ffffffffc090090f>] zfsdev_ioctl+0x60f/0x6c0 [zfs]
[<ffffffff8f058dc1>] do_vfs_ioctl+0xa1/0x5f0
[<ffffffff8f059389>] SyS_ioctl+0x79/0x90
[<ffffffff8f6d3ebb>] entry_SYSCALL_64_fastpath+0x1e/0xad
[<ffffffffffffffff>] 0xffffffffffffffff
{code}
This is happening because when the test is destroying datasets and then destroying the pool, it is forcibly evicting all the dbufs in each objset_t. The dbuf_evict scavenger thread is also trying to evict the oldest dbufs, and introducing a data race for the per file (dnode) list of dbufs.
Proposed solution is the enforce mutual exclusion between the objset eviction when ZFS is done with a dataset and the dbuf_evict process which is evicting old dbufs.

Reply:
Thanks for providing the stack traces, that helps.  But it's still not clear to me why the logic in `dbuf_destroy()` and `dnode_evict_dbufs()` fail to correctly handle this race.  The existing locking looks sufficient, but clearly it's not.

The thread calling `dbuf_evict_one()->dbuf_destroy()` must have set `db->db_state = DB_EVICTING` before dropping the `db->db_mtx`.  This should prevent `dnode_evict_dbufs()` from taking the first branch of the `if (db->db_state != DB_EVICTING && ...)` conditional, thereby avoiding the call to `avl_insert_here()` and the panic.  Furthermore, all updates to the ` dn->dn_dbufs` are correctly done under the `dn->dn_dbufs_mtx`.  What am I missing?

Reply:
There is also the zpool destroy happening.  I'm currently trying to figure out if dn_dbufs is getting wiped out by dnode_destroy prematurely.  I thought that only happened when the user buf gets evicted.

Reply:
A little more detail on the failure during dnode_evict_dbufs()
```May  9 15:40:48 ub16fresh kernel: [13474.205129] VERIFY(diff > 0 ? child == 1 : child == 0) failed
May  9 15:40:48 ub16fresh kernel: [13474.205132] PANIC at avl.c:620:avl_insert_here()
```

Reply:
So we have a zvol testpool/testvol which we snap (testpool/testvol@testsnap) and clone: zfs clone testpool/testvol@testsnap testpool/testclone1.  During test cleanup we destroy this zvol with zfs destroy -Rf testpool/testvol. What I see is that the dnode_t representing the zvol is being evicted by multiple threads.
{code}
May 20 16:28:10 ub16fresh kernel: [ 5735.560223] NOTICE: evicting os ffff97dcf8789000 118 spa ffff97dceac8c000 dn ffff97dcd1519200 1 thread 9768
May 20 16:28:10 ub16fresh kernel: [ 5735.560225] CPU: 0 PID: 9768 Comm: z_zvol Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 20 16:28:10 ub16fresh kernel: [ 5735.560226] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 20 16:28:10 ub16fresh kernel: [ 5735.560226] Call Trace:
May 20 16:28:10 ub16fresh kernel: [ 5735.560231]  dump_stack+0x63/0x90
May 20 16:28:10 ub16fresh kernel: [ 5735.560270]  dmu_objset_evict_dbufs+0x2a9/0x2b0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560339]  zvol_shutdown_zv+0xa8/0x110 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560463]  zvol_last_close+0x3a/0xb0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560508]  zvol_release+0x172/0x3d0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560511]  __blkdev_put+0x235/0x290
May 20 16:28:10 ub16fresh kernel: [ 5735.560513]  ? unlock_new_inode+0x53/0x80
May 20 16:28:10 ub16fresh kernel: [ 5735.560514]  blkdev_put+0x4c/0x110
May 20 16:28:10 ub16fresh kernel: [ 5735.560516]  device_add_disk+0x409/0x480
May 20 16:28:10 ub16fresh kernel: [ 5735.560558]  ? __list_add+0x17/0x40 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560600]  add_disk+0xe/0x10 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560642]  zvol_create_minor_impl+0x48f/0x560 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560672]  ? kfree+0x162/0x170
May 20 16:28:10 ub16fresh kernel: [ 5735.560777]  zvol_create_minors_impl+0x12b/0x1c0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560780]  ? __schedule+0x23a/0x700
May 20 16:28:10 ub16fresh kernel: [ 5735.560927]  zvol_task_cb+0x67/0xe0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.560942]  taskq_thread+0x262/0x4d0 [spl]
May 20 16:28:10 ub16fresh kernel: [ 5735.560944]  ? wake_up_q+0x70/0x70
May 20 16:28:10 ub16fresh kernel: [ 5735.560948]  kthread+0x109/0x140
May 20 16:28:10 ub16fresh kernel: [ 5735.560952]  ? taskq_thread_spawn+0x50/0x50 [spl]
May 20 16:28:10 ub16fresh kernel: [ 5735.560954]  ? kthread_create_on_node+0x60/0x60
May 20 16:28:10 ub16fresh kernel: [ 5735.560957]  ret_from_fork+0x2c/0x40
May 20 16:28:10 ub16fresh kernel: [ 5735.578902] NOTICE: evicting os ffff97dcf56fd000 118 spa ffff97dceac8c000 dn ffff97dcd1519200 1 thread 10629
May 20 16:28:10 ub16fresh kernel: [ 5735.578904] CPU: 0 PID: 10629 Comm: systemd-udevd Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 20 16:28:10 ub16fresh kernel: [ 5735.578905] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 20 16:28:10 ub16fresh kernel: [ 5735.578906] Call Trace:
May 20 16:28:10 ub16fresh kernel: [ 5735.578911]  dump_stack+0x63/0x90
May 20 16:28:10 ub16fresh kernel: [ 5735.578951]  dmu_objset_evict_dbufs+0x2a9/0x2b0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.578992]  zvol_shutdown_zv+0xa8/0x110 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.579032]  zvol_last_close+0x3a/0xb0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.579072]  zvol_release+0x172/0x3d0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.579074]  __blkdev_put+0x235/0x290
May 20 16:28:10 ub16fresh kernel: [ 5735.579075]  blkdev_put+0x4c/0x110
May 20 16:28:10 ub16fresh kernel: [ 5735.579076]  blkdev_close+0x34/0x70
May 20 16:28:10 ub16fresh kernel: [ 5735.579077]  __fput+0xe7/0x220
May 20 16:28:10 ub16fresh kernel: [ 5735.579078]  ____fput+0xe/0x10
May 20 16:28:10 ub16fresh kernel: [ 5735.579081]  task_work_run+0x7e/0xa0
May 20 16:28:10 ub16fresh kernel: [ 5735.579083]  exit_to_usermode_loop+0xaa/0xb0
May 20 16:28:10 ub16fresh kernel: [ 5735.579084]  syscall_return_slowpath+0x59/0x60
May 20 16:28:10 ub16fresh kernel: [ 5735.579087]  entry_SYSCALL_64_fastpath+0xab/0xad
May 20 16:28:10 ub16fresh kernel: [ 5735.579088] RIP: 0033:0x7f3ca2cf3560
May 20 16:28:10 ub16fresh kernel: [ 5735.579088] RSP: 002b:00007ffc47b75948 EFLAGS: 00000246 ORIG_RAX: 0000000000000003
May 20 16:28:10 ub16fresh kernel: [ 5735.579089] RAX: 0000000000000000 RBX: 0000000000000007 RCX: 00007f3ca2cf3560
May 20 16:28:10 ub16fresh kernel: [ 5735.579090] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000007
May 20 16:28:10 ub16fresh kernel: [ 5735.579090] RBP: 00007f3ca3b8d708 R08: 00005647d5200370 R09: 0000000000000000
May 20 16:28:10 ub16fresh kernel: [ 5735.579091] R10: 00005647d434f899 R11: 0000000000000246 R12: 000000000000000b
May 20 16:28:10 ub16fresh kernel: [ 5735.579091] R13: 0000000000000003 R14: 000000000000000e R15: 0000000000000000
May 20 16:28:10 ub16fresh kernel: [ 5735.607600] NOTICE: evicting os ffff97dcf8789000 118 spa ffff97dceac8c000 dn ffff97dcd1519200 1 thread 10635
May 20 16:28:10 ub16fresh kernel: [ 5735.607603] CPU: 0 PID: 10635 Comm: systemd-udevd Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 20 16:28:10 ub16fresh kernel: [ 5735.607603] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 20 16:28:10 ub16fresh kernel: [ 5735.607604] Call Trace:
May 20 16:28:10 ub16fresh kernel: [ 5735.607609]  dump_stack+0x63/0x90
May 20 16:28:10 ub16fresh kernel: [ 5735.607652]  dmu_objset_evict_dbufs+0x2a9/0x2b0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.607697]  zvol_shutdown_zv+0xa8/0x110 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.607740]  zvol_last_close+0x3a/0xb0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.607784]  zvol_release+0x172/0x3d0 [zfs]
May 20 16:28:10 ub16fresh kernel: [ 5735.607786]  __blkdev_put+0x235/0x290
May 20 16:28:10 ub16fresh kernel: [ 5735.607787]  blkdev_put+0x4c/0x110
May 20 16:28:10 ub16fresh kernel: [ 5735.607788]  blkdev_close+0x34/0x70
May 20 16:28:10 ub16fresh kernel: [ 5735.607790]  __fput+0xe7/0x220
May 20 16:28:10 ub16fresh kernel: [ 5735.607791]  ____fput+0xe/0x10
May 20 16:28:10 ub16fresh kernel: [ 5735.607793]  task_work_run+0x7e/0xa0
May 20 16:28:10 ub16fresh kernel: [ 5735.607795]  exit_to_usermode_loop+0xaa/0xb0
May 20 16:28:10 ub16fresh kernel: [ 5735.607796]  syscall_return_slowpath+0x59/0x60
May 20 16:28:10 ub16fresh kernel: [ 5735.607799]  entry_SYSCALL_64_fastpath+0xab/0xad
{code}
You would think the DMU locking (dn->dv_dbufs_mtx) would still sort this out but we seem to be abusing the eviction capability here.  My fix likely works because it allows only one thread at a time to evict the objset_t.

Reply:
Here's another example where systemd-udevd and dbu_evict are both evicting the same objset and dnode.
```
May 21 10:37:25 ub16fresh kernel: [ 3065.701274] NOTICE: zv ffff923e3bbac800 os 118 ffff923e19c0f800 open 3 dn ffff923e39218d98 name testpool/testclone1 pid 11190
May 21 10:37:25 ub16fresh kernel: [ 3065.702550] NOTICE: zv ffff923e3bbac800 os 118 ffff923e19c0f800 open 3 dn ffff923e39218d98 name testpool/testclone1 pid 11191
May 21 10:37:25 ub16fresh kernel: [ 3065.726265] NOTICE: zv ffff923e3bbac800 os 118 ffff923e19c0f800 open 2 dn ffff923e39218d98 name testpool/testclone1 pid 11117
May 21 10:37:25 ub16fresh kernel: [ 3065.748293] NOTICE: zv ffff923e3bbac800 os 118 ffff923e19c0f800 open 1 dn ffff923e39218d98 name testpool/testclone1 pid 11160
May 21 10:37:25 ub16fresh kernel: [ 3065.823158] NOTICE: evicting os ffff923e19c0f800 118 spa ffff923e0d820000 dn ffff923e39218d98 1 thread 11160
May 21 10:37:25 ub16fresh kernel: [ 3065.823162] CPU: 0 PID: 11160 Comm: systemd-udevd Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 21 10:37:25 ub16fresh kernel: [ 3065.823163] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 21 10:37:25 ub16fresh kernel: [ 3065.823164] Call Trace:
May 21 10:37:25 ub16fresh kernel: [ 3065.823170]  dump_stack+0x63/0x90
May 21 10:37:25 ub16fresh kernel: [ 3065.823223]  dmu_objset_evict_dbufs+0x2a9/0x2b0 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.823278]  zvol_shutdown_zv+0x92/0xc0 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.823327]  zvol_last_close+0x3a/0xb0 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.823375]  zvol_release+0x18c/0x470 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.823377]  __blkdev_put+0x235/0x290
May 21 10:37:25 ub16fresh kernel: [ 3065.823378]  blkdev_put+0x4c/0x110
May 21 10:37:25 ub16fresh kernel: [ 3065.823379]  blkdev_close+0x34/0x70
May 21 10:37:25 ub16fresh kernel: [ 3065.823381]  __fput+0xe7/0x220
May 21 10:37:25 ub16fresh kernel: [ 3065.823382]  ____fput+0xe/0x10
May 21 10:37:25 ub16fresh kernel: [ 3065.823385]  task_work_run+0x7e/0xa0
May 21 10:37:25 ub16fresh kernel: [ 3065.823387]  exit_to_usermode_loop+0xaa/0xb0
May 21 10:37:25 ub16fresh kernel: [ 3065.823388]  syscall_return_slowpath+0x59/0x60
May 21 10:37:25 ub16fresh kernel: [ 3065.823391]  entry_SYSCALL_64_fastpath+0xab/0xad
May 21 10:37:25 ub16fresh kernel: [ 3065.823392] RIP: 0033:0x7efdba5b3560
May 21 10:37:25 ub16fresh kernel: [ 3065.823393] RSP: 002b:00007ffd34f725a8 EFLAGS: 00000246 ORIG_RAX: 0000000000000003
May 21 10:37:25 ub16fresh kernel: [ 3065.823395] RAX: 0000000000000000 RBX: 0000000000000007 RCX: 00007efdba5b3560
May 21 10:37:25 ub16fresh kernel: [ 3065.823395] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000007
May 21 10:37:25 ub16fresh kernel: [ 3065.823396] RBP: 00007efdbb44d708 R08: 0000555de82eecf0 R09: 0000000000000000
May 21 10:37:25 ub16fresh kernel: [ 3065.823396] R10: 0000555de769d899 R11: 0000000000000246 R12: 000000000000000b
May 21 10:37:25 ub16fresh kernel: [ 3065.823397] R13: 0000000000000003 R14: 000000000000000e R15: 0000000000000000
May 21 10:37:25 ub16fresh kernel: [ 3065.838130] NOTICE: evicting os ffff923e19c0f800 118 spa ffff923e0d820000 dn ffff923e39218d98 1 thread 16054
May 21 10:37:25 ub16fresh kernel: [ 3065.838133] CPU: 0 PID: 16054 Comm: dbu_evict Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 21 10:37:25 ub16fresh kernel: [ 3065.838134] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 21 10:37:25 ub16fresh kernel: [ 3065.838135] Call Trace:
May 21 10:37:25 ub16fresh kernel: [ 3065.838141]  dump_stack+0x63/0x90
May 21 10:37:25 ub16fresh kernel: [ 3065.838194]  dmu_objset_evict_dbufs+0x2a9/0x2b0 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.838235]  dmu_objset_evict+0x115/0x260 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.838276]  dsl_dataset_evict_async+0x3b/0x280 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.838284]  taskq_thread+0x262/0x4d0 [spl]
May 21 10:37:25 ub16fresh kernel: [ 3065.838287]  ? wake_up_q+0x70/0x70
May 21 10:37:25 ub16fresh kernel: [ 3065.838328]  ? dsl_dataset_feature_is_active+0x20/0x20 [zfs]
May 21 10:37:25 ub16fresh kernel: [ 3065.838331]  kthread+0x109/0x140
May 21 10:37:25 ub16fresh kernel: [ 3065.838337]  ? taskq_thread_spawn+0x50/0x50 [spl]
May 21 10:37:25 ub16fresh kernel: [ 3065.838339]  ? kthread_create_on_node+0x60/0x60
May 21 10:37:25 ub16fresh kernel: [ 3065.838342]  ret_from_fork+0x2c/0x40
```

Reply:
Hi @behlendorf I think I finally have some info about the root cause that is worth sharing.  In the zfs test script final steps, we create a zvol, fail the snap -r as expected, and go right into destroying datasets in the cleanup function.  What I see is that dbuf_evict_one code is picking off a zvol dbuf before zvol_create_minor is finished.  With the db_state of DB_EVICTING, continuation of the zvol_create_minor code holds a new dbuf into dn_dbufs with the same bookmark and state of DB_CACHED.  During objset eviction, this arrangement seems to cause issues with the avl_insert_here insertion sort (dbuf_compare) resulting in the panic.  See below:

{code}
First dbuf hold ffff8bb449141370

May 30 11:34:33 ub16fresh kernel: [ 5152.106910] NOTICE: 118:0:0:0 p = ffff8bb449141370
May 30 11:34:33 ub16fresh kernel: [ 5152.106912] CPU: 0 PID: 17393 Comm: z_zvol Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 30 11:34:33 ub16fresh kernel: [ 5152.106912] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 30 11:34:33 ub16fresh kernel: [ 5152.106912] Call Trace:
May 30 11:34:33 ub16fresh kernel: [ 5152.106914]  dump_stack+0x63/0x90
May 30 11:34:33 ub16fresh kernel: [ 5152.106938]  dbuf_hold_level+0xac/0xb0 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.106960]  dbuf_hold+0x16/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.106986]  dnode_hold_impl+0x30f/0xee0 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107018]  ? queued_spin_unlock+0xb/0x10 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107050]  ? zfs_refcount_remove+0x16/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107085]  ? spa_close+0x3b/0x80 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107111]  dnode_hold+0x1b/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107134]  dmu_object_info+0x30/0x80 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107173]  zvol_create_minor_impl+0x1fa/0x560 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107175]  ? kfree+0x162/0x170
May 30 11:34:33 ub16fresh kernel: [ 5152.107213]  ? __list_del_entry+0x1d/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.107251]  zvol_create_minors_impl+0x12b/0x1c0 [zfs]


First evict ffff8bb449141370

May 30 11:34:33 ub16fresh kernel: [ 5152.112256] NOTICE: changing db ffff8bb449141370 state to DB_EVICTING
May 30 11:34:33 ub16fresh kernel: [ 5152.112257] CPU: 0 PID: 12900 Comm: dbuf_evict Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 30 11:34:33 ub16fresh kernel: [ 5152.112258] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 30 11:34:33 ub16fresh kernel: [ 5152.112258] Call Trace:
May 30 11:34:33 ub16fresh kernel: [ 5152.112260]  dump_stack+0x63/0x90
May 30 11:34:33 ub16fresh kernel: [ 5152.112287]  dbuf_destroy+0x66d/0x680 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.112323]  ? zfs_refcount_remove_many+0x12f/0x200 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.112347]  dbuf_evict_one+0x178/0x300 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.112371]  dbuf_evict_thread+0x141/0x270 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.112393]  ? dbuf_evict_one+0x300/0x300 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.112398]  thread_generic_wrapper+0x7b/0xc0 [spl]
May 30 11:34:33 ub16fresh kernel: [ 5152.112400]  kthread+0x109/0x140
May 30 11:34:33 ub16fresh kernel: [ 5152.112404]  ? IS_ERR+0x10/0x10 [spl]
May 30 11:34:33 ub16fresh kernel: [ 5152.112405]  ? kthread_create_on_node+0x60/0x60
May 30 11:34:33 ub16fresh kernel: [ 5152.112407]  ret_from_fork+0x2c/0x40


Hold new dbuf ffff8bb44e864c08  after evict of ffff8bb449141370

May 30 11:34:33 ub16fresh kernel: [ 5152.114029] NOTICE: 118:0:0:0 p = ffff8bb44e864c08
May 30 11:34:33 ub16fresh kernel: [ 5152.114030] CPU: 0 PID: 17393 Comm: z_zvol Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 30 11:34:33 ub16fresh kernel: [ 5152.114031] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 30 11:34:33 ub16fresh kernel: [ 5152.114031] Call Trace:
May 30 11:34:33 ub16fresh kernel: [ 5152.114034]  dump_stack+0x63/0x90
May 30 11:34:33 ub16fresh kernel: [ 5152.114062]  dbuf_hold_level+0xac/0xb0 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114086]  dbuf_hold+0x16/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114115]  dnode_hold_impl+0x30f/0xee0 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114153]  ? do_raw_spin_unlock+0x9/0x10 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114191]  ? __raw_spin_unlock+0x9/0x10 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114229]  ? txg_wait_synced+0x166/0x210 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114231]  ? vsnprintf+0x460/0x4c0
May 30 11:34:33 ub16fresh kernel: [ 5152.114260]  dnode_hold+0x1b/0x20 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114285]  dmu_prefetch+0x8b/0x260 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114327]  zvol_create_minor_impl+0x4f6/0x560 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.114329]  ? kfree+0x162/0x170
May 30 11:34:33 ub16fresh kernel: [ 5152.114371]  zvol_create_minors_impl+0x12b/0x1c0 [zfs]

Panic with ffff8bb449141370 and ffff8bb44e864c08 both in dn_dbufs with same bookmark, different db_state

May 30 11:34:33 ub16fresh kernel: [ 5152.117619] DUMPING DBUFS FOR OS = 118, OBJ = 0 PID = 12898
May 30 11:34:33 ub16fresh kernel: [ 5152.117620]        dbuf 0: zb = 118:0:0:0 state = 5 p = ffff8bb449141370
May 30 11:34:33 ub16fresh kernel: [ 5152.117621]        dbuf 1: zb = 118:0:0:0 state = 4 p = ffff8bb44e864c08
May 30 11:34:33 ub16fresh kernel: [ 5152.117621]        dbuf 2: zb = 118:0:1:0 state = 4 p = ffff8bb449141000
May 30 11:34:33 ub16fresh kernel: [ 5152.117622]        dbuf 3: zb = 118:0:2:0 state = 4 p = ffff8bb4491411b8
May 30 11:34:33 ub16fresh kernel: [ 5152.117623]        dbuf 4: zb = 118:0:3:0 state = 4 p = ffff8bb449141dc0
May 30 11:34:33 ub16fresh kernel: [ 5152.117623]        dbuf 5: zb = 118:0:4:0 state = 4 p = ffff8bb449141a50
May 30 11:34:33 ub16fresh kernel: [ 5152.117624]        dbuf 6: zb = 118:0:5:0 state = 4 p = ffff8bb4491416e0
May 30 11:34:33 ub16fresh kernel: [ 5152.117624] ----------------------------------------
May 30 11:34:33 ub16fresh kernel: [ 5152.117625] VERIFY(diff > 0 ? child == 1 : child == 0) failed
May 30 11:34:33 ub16fresh kernel: [ 5152.117627] PANIC at avl.c:621:avl_insert_here()
May 30 11:34:33 ub16fresh kernel: [ 5152.117628] Showing stack for process 12898
May 30 11:34:33 ub16fresh kernel: [ 5152.117629] CPU: 0 PID: 12898 Comm: dbu_evict Tainted: P           OE   4.10.0-28-generic #32~16.04.2-Ubuntu
May 30 11:34:33 ub16fresh kernel: [ 5152.117629] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
May 30 11:34:33 ub16fresh kernel: [ 5152.117630] Call Trace:
May 30 11:34:33 ub16fresh kernel: [ 5152.117631]  dump_stack+0x63/0x90
May 30 11:34:33 ub16fresh kernel: [ 5152.117634]  spl_dumpstack+0x29/0x30 [spl]
May 30 11:34:33 ub16fresh kernel: [ 5152.117636]  spl_panic+0xc5/0x100 [spl]
May 30 11:34:33 ub16fresh kernel: [ 5152.117638]  ? vprintk_emit+0x2ea/0x4b0
May 30 11:34:33 ub16fresh kernel: [ 5152.117639]  ? vprintk_default+0x29/0x50
May 30 11:34:33 ub16fresh kernel: [ 5152.117640]  ? printk+0x57/0x73
May 30 11:34:33 ub16fresh kernel: [ 5152.117642]  avl_insert_here+0x43f/0x450 [zavl]
May 30 11:34:33 ub16fresh kernel: [ 5152.117666]  dnode_evict_dbufs+0x115/0x3e0 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.117690]  dmu_objset_evict_dbufs+0x136/0x220 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.117713]  dmu_objset_evict+0xdc/0x220 [zfs]
May 30 11:34:33 ub16fresh kernel: [ 5152.117738]  dsl_dataset_evict_async+0x3b/0x280 [zfs]

{code}

The fix I have making only one thread at a time evict an objset_t works here but maybe there's a better solution?

Reply:
nice finding out what was going on, and the concept seems reasonable.